File: app/src/main/java/jp/juggler/subwaytooter/table/MediaShown.java
package jp.juggler.subwaytooter.table;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.support.annotation.NonNull;

import jp.juggler.subwaytooter.App1;
import jp.juggler.subwaytooter.api.entity.TootStatusLike;
import jp.juggler.subwaytooter.util.LogCategory;

public class MediaShown {
	private static final LogCategory log = new LogCategory( "MediaShown" );
	
	private static final String table = "media_shown";
	private static final String COL_HOST = "h";
	private static final String COL_STATUS_ID = "si";
	private static final String COL_SHOWN = "sh";
	private static final String COL_TIME_SAVE = "time_save";
	
	public static void onDBCreate( SQLiteDatabase db ){
		log.d("onDBCreate!");
		db.execSQL(
			"create table if not exists " + table
				+ "(_id INTEGER PRIMARY KEY"
				+ ",h text not null"
				+ ",si integer not null"
				+ ",sh integer not null"
				+ ",time_save integer default 0"
				+ ")"
		);
		db.execSQL(
			"create unique index if not exists " + table + "_status_id on " + table + "(h,si)"
		);
	}
	
	public static void onDBUpgrade( SQLiteDatabase db, int oldVersion, int newVersion ){
		if( oldVersion < 5 && newVersion >= 5 ){
			db.execSQL( "drop table if exists "+table);
			onDBCreate( db );
		}
	}
	
	private static final String[] projection_shown = new String[]{COL_SHOWN};
	
	public static boolean isShown( @NonNull TootStatusLike status , boolean default_value ){
		try{
			Cursor cursor = App1.getDB().query( table, projection_shown, "h=? and si=?", new String[]{ status.getHostAccessOrOriginal(), Long.toString(status.getIdAccessOrOriginal()) }, null, null, null );
			try{
				if( cursor.moveToFirst() ){
					return ( 0 != cursor.getInt( cursor.getColumnIndex( COL_SHOWN ) ) );
				}
			}finally{
				cursor.close();
			}
		}catch( Throwable ex ){
			log.e( ex, "load failed." );
		}
		return default_value ;
	}
	
	public static void save( @NonNull TootStatusLike status ,boolean is_shown ){
		try{
			long now = System.currentTimeMillis();
			
			ContentValues cv = new ContentValues();
			cv.put( COL_HOST, status.getHostAccessOrOriginal() );
			cv.put( COL_STATUS_ID, status.getIdAccessOrOriginal() );
			cv.put( COL_SHOWN, is_shown ? 1:0 );
			cv.put( COL_TIME_SAVE, now );
			App1.getDB().replace( table, null, cv );
			
			// 古いデータを掃除する
			long expire = now - 86400000L * 365;
			App1.getDB().delete( table,COL_TIME_SAVE+"<?",new String[]{Long.toString(expire)});
			
			
		}catch( Throwable ex ){
			log.e( ex, "save failed." );
		}
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/table/MediaShown.kt
package jp.juggler.subwaytooter.table

import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import android.provider.BaseColumns
import jp.juggler.subwaytooter.api.entity.TootStatus
import jp.juggler.util.data.MetaColumns
import jp.juggler.util.data.TableCompanion
import jp.juggler.util.data.getBoolean
import jp.juggler.util.log.LogCategory

class MediaShown private constructor() {
    companion object : TableCompanion {
        private val log = LogCategory("MediaShown")

        override val table = "media_shown_misskey"
        private const val COL_ID = BaseColumns._ID
        private const val COL_HOST = "h"
        private const val COL_STATUS_ID = "si"
        private const val COL_SHOWN = "sh"
        private const val COL_TIME_SAVE = "time_save"

        val columnList = MetaColumns(table, 30).apply {
            column(0, COL_ID, "INTEGER PRIMARY KEY")
            column(0, COL_HOST, "")
            column(0, COL_STATUS_ID, "")
            column(0, COL_SHOWN, "")
            column(0, COL_TIME_SAVE, "")
            deleteBeforeCreate = true
            createExtra = {
                arrayOf(
                    "create unique index if not exists ${table}_status_id on $table($COL_HOST,$COL_STATUS_ID)",
                    "create index if not exists ${table}_time_save on $table($COL_TIME_SAVE)",
                )
            }
        }

        override fun onDBCreate(db: SQLiteDatabase) =
            columnList.onDBCreate(db)

        override fun onDBUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
            columnList.onDBUpgrade(db, oldVersion, newVersion)

            // 特定バージョンを交差したらテーブルを作り直す
            if (oldVersion < 30 && newVersion >= 30) {
                columnList.onDBCreate(db)
            }
        }

        private val projection_shown = arrayOf(COL_SHOWN)
    }

    class Access(val db: SQLiteDatabase) {

        fun deleteOld(now: Long) {
            try {
                val expire = now - 86400000L * 365
                db.delete(table, "$COL_TIME_SAVE<?", arrayOf(expire.toString()))
            } catch (ex: Throwable) {
                log.e(ex, "deleteOld failed.")
            }

            // 旧型式のテーブルも古い項目の削除だけ行う
            try {
                val table = "media_shown"
                val COL_TIME_SAVE = "time_save"
                val expire = now - 86400000L * 365
                db.delete(table, "$COL_TIME_SAVE<?", arrayOf(expire.toString()))
            } catch (ignored: Throwable) {
            }
        }

        private fun saveImpl(host: String, id: String, isShown: Boolean) {
            try {
                ContentValues().apply {
                    put(COL_HOST, host)
                    put(COL_STATUS_ID, id)
                    put(COL_SHOWN, isShown)
                    put(COL_TIME_SAVE, System.currentTimeMillis())
                }.let { db.replace(table, null, it) }
            } catch (ex: Throwable) {
                log.e(ex, "saveImpl failed.")
            }
        }

        private fun isShownImpl(host: String, id: String, defaultValue: Boolean): Boolean {
            try {
                db.query(
                    table,
                    projection_shown,
                    "h=? and si=?",
                    arrayOf(host, id),
                    null,
                    null,
                    null
                ).use { cursor ->
                    if (cursor.moveToFirst()) return cursor.getBoolean(COL_SHOWN)
                }
            } catch (ex: Throwable) {
                log.e(ex, "isShownImpl failed.")
            }
            return defaultValue
        }

        fun save(uri: String, isShown: Boolean) =
            saveImpl(uri, uri, isShown)

        fun isShown(uri: String, defaultValue: Boolean) =
            isShownImpl(uri, uri, defaultValue)

        fun save(status: TootStatus, isShown: Boolean) =
            saveImpl(status.hostAccessOrOriginal.ascii, status.id.toString(), isShown)

        fun isShown(status: TootStatus, defaultValue: Boolean) =
            isShownImpl(status.hostAccessOrOriginal.ascii, status.id.toString(), defaultValue)
    }
}
